### PR TITLE
pkgs/sagemath*: Exclude all__*.py files of other distributions

### PR DESCRIPTION
**Pull request by mkoeppe**
Fixes https://github.com/sagemath/sage/pull/35661#issuecomment-1659211122

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by kiwifb** [comment]
I may be saying stupid here, how will the PR in its current state prevent `sagemath-standard` to ship, say, `sage/graphs/bliss.pyx`? Or have the design of the thing completely wrong?

**Comment by mkoeppe** [comment]
Sorry, I missed the part about bliss.pyx in your report. I'll take a look.

**Comment by kiwifb** [comment]
Similarly, bits belonging to meataxe and sirocco were shipped as part of sagemath-standard. I cannot say for the other two but it may be the case as well.

**Comment by kiwifb** [comment]
I have just inspected the "Manifest.in" for sagemath-standard and it does include an exclude line for bliss
```
# Exclude extension modules shipped by optional packages
exclude sage/graphs/bliss.pyx
```
but none for the other optional packages. But even so it is there, I still have bliss.pyx shipped currently. I am not sure why it is ignored.

**Comment by kiwifb** [comment]
OK, `MANIFEST.in` is used for building sdits. Which means that I use the sdist for sagemath-standard-10.0, `bliss.pyx` is not included and everything is fine. But `MANIFEST.in` is not used when building wheels. My current install process produces a wheel and install it (for both master and develop) but because develop is a git clone, it includes `bliss.pyx` and other files that would otherwise be excluded in a sdist. It seems the only way to exclude a file from a wheel is to explicitly exclude it with a
```
find_packages(exclude=[...])
```
pattern. Which I find problematic from a maintenance point of view but probably doable.

**Comment by kiwifb** [comment]
We may want `exclude_package_data` https://setuptools.pypa.io/en/latest/userguide/datafiles.html#exclude-package-data

**Comment by mkoeppe** [comment]
I think we need to build the wheels via the sdists so that filtering works properly.

I'm already doing this in #35095 for most distributions.
I'll look into this next.

**Comment by mkoeppe** [comment]
There is unfortunately a limitation of setuptools: If I want to declare package data, I need to first declare a package. But this then leads to include all Python files in that package as source files.

This will all go away when we move from setuptools to something better (#34630). In the meantime, building wheel via sdist is the solution. This is also the default way how the "build" tools builds wheels.

**Comment by kiwifb** [comment]
Well it is a bit of an issue when you start from a git clone. The process right now would be
* git clone
* prepare sdist
* use sdist to build package
The prepare sdist step is unusual and not well catered for in an ebuild environment. The normal solution is to pre-package the sdist but that nullify any benefits from the cloning ability to stay on top of what is happening.

But it remains that the current `MANIFEST.in` for sage needs to exclude files for meataxe, sirocco and the other optional packages so they do not get in the sdist. Let's deal with this particular problem here.

**Comment by mkoeppe** [comment]
> * The prepare sdist step is unusual and not well catered for in an ebuild environment.

You mean ebuild insists on using `setup.py bdist_wheel` instead of using modern tools such as https://pypa-build.readthedocs.io/en/latest/#python--m-build ?

**Comment by kiwifb** [comment]
> > * The prepare sdist step is unusual and not well catered for in an ebuild environment.
> 
> You mean ebuild insists on using `setup.py bdist_wheel` instead of using modern tools such as https://pypa-build.readthedocs.io/en/latest/#python--m-build ?

Out of my hands at this stage. I use the standard boiler plate I am provided with - which mostly expect a sdist or sdist like thing. Part of the problem is that process takes precedence over whatever the package maintainer would want to do. 

Philosophically, I consider the whole thing as python wanting to have nothing to do with downstream packaging and not supporting it. This is the way we package things in python, deal with it. A current that seems to have gain prevalence in a number of ecosystems in the last decade for good or bad. I do consider it a mixed bag myself, but coming from a rolling release distribution like Gentoo, I appreciate wanting to escape the normal binary distribution model. Using binary python package from a traditional distribution does put serious limits on the user.

**Comment by kiwifb** [comment]
OK, I can make things work in Gentoo so long as `MANIFEST.in` is correct. I can leverage that. I am adding stuff for sirocco, meataxe, mcqd, coxeter3 and tdlib. What about all those "all__*` empty files? Should they also be excluded from sagemath-standard? It looks like they are all installed by default in develop.

File: pkgs/sagemath-bliss/MANIFEST.in [hunk 1]
```diff
@@ -1,12 +1,13 @@
-global-include all__sagemath_bliss.py
-
 include VERSION.txt
 
 graft sage/graphs/bliss_cpp
```
Comment (kiwifb) [inline pkgs/sagemath-bliss/MANIFEST.in:3]: What is this file? Is it supposed to be `bliss.cpp`, a generated file from `bliss.pyx`?

**Comment by mkoeppe** [comment]
> What about all those "all__*` empty files? Should they also be excluded from sagemath-standard?

Yes, done in a8d6ef0

File: pkgs/sagemath-bliss/MANIFEST.in (hunk 1 above)
Comment (mkoeppe) [reply pkgs/sagemath-bliss/MANIFEST.in:3]: It's a directory that contains `bliss_find_automorphisms.h` since a recent update

Comment (kiwifb) [reply pkgs/sagemath-bliss/MANIFEST.in:3]: OK, that is stuff from the bliss update, I thought that was merged in develop, that explains why I cannot see it.

**Comment by kiwifb** [comment]
> > What about all those "all__*` empty files? Should they also be excluded from sagemath-standard?
> 
> Yes, done in [a8d6ef0](https://github.com/sagemath/sage/commit/a8d6ef0720223ae7e3ce81590581db1bb2b3d586)

Thanks and now I have to refresh and restart my edits on sagemath-standard's MANIFEST.in.

**Comment by mkoeppe** [comment]
Just pushed another fix of files that shouldn't be shipped by sagemath-standard any more.

**Comment by kiwifb** [comment]
That take care of what I was working on. For tdlib I think there should be
a pyx file as well as the .cpp.

File: src/MANIFEST.in [hunk 2]
```diff
@@ -45,7 +44,22 @@ include sage/geometry/triangulation/data.cc
...
 
 # Exclude extension modules shipped by optional packages
-exclude sage/graphs/bliss.pyx
+exclude sage/graphs/bliss.p*
+prune sage/graphs/bliss_cpp
+prune sage/libs/coxeter3
+exclude sage/graphs/mcqd.p*
+exclude sage/libs/meataxe.p*
+exclude sage/matrix/matrix_gfpn_dense.p*
+exclude sage/graphs/graph_decompositions/sage_tdlib.cpp
```
Comment (kiwifb) [inline src/MANIFEST.in]: This file should already be excluded by `global-exclude *.cpp` on line 15. And its explicit inclusion in an earlier section is also removed. So, I do not think we need to keep this line.

**Comment by kiwifb** [comment]
Apart from the last question about that .cpp file, I tested that this branch worked properly at least for bliss, meataxe and sirocco. I think we are able to ship it ASAP.

**Comment by mkoeppe** [comment]
I thought I took care of the tdlib things already in cf02a42

**Comment by mkoeppe** [comment]
I'll postpone the switch to using python-build to a follow-up PR

**Comment by kiwifb** [comment]
> I thought I took care of the tdlib things already in [cf02a42](https://github.com/sagemath/sage/commit/cf02a4256ee8e630ec444452d1dfdd0dac7382cf)

Did you read my review comment? It works as is, but the line excluding `sage_tdlib.cpp` that comes from the previous commit is redundant, removing the corresponding include line was sufficient to make sure it is not shipped.

**Comment by mkoeppe** [comment]
> Did you read my review comment?

I don't see a comment, maybe you didn't "submit" the review?

>  It works as is, but the line excluding `sage_tdlib.cpp` that comes from the previous commit is redundant, removing the corresponding include line was sufficient to make sure it is not shipped.

That's true, I'll remove it again.

**Review by kiwifb — APPROVED** [review]
Yes I must have "forgotten" to submit. It should go now.

**Comment by mkoeppe** [comment]
May I set to positive review then?

**Comment by kiwifb** [comment]
Sorry forgot to change the label.

**Comment by mkoeppe** [comment]
Thank you!

**Comment by kiwifb** [comment]
Now that rc0 is out and I have run proper doctests on sage-on-gentoo, I have noticed unintended consequences that I missed before.
```
sage -t --long --random-seed=118675773581622755375337172702839208422 /usr/lib/python3.11/site-packages/sage_setup/find.py
**********************************************************************
File "/usr/lib/python3.11/site-packages/sage_setup/find.py", line 117, in sage_setup.find.?
Failed example:
    find_python_sources(SAGE_SRC, distributions=['sagemath-tdlib'])
Expected:
    ([], [], [<setuptools.extension.Extension('sage.graphs.graph_decompositions.tdlib')...>])
Got:
    ([], [], [])
**********************************************************************
File "/usr/lib/python3.11/site-packages/sage_setup/find.py", line 196, in sage_setup.find.filter_cython_sources
Failed example:
    any(f.endswith('sage/graphs/graph_decompositions/tdlib.pyx') for f in cython_modules)
Expected:
    True
Got:
    False
```
and
```
sage -t --long --random-seed=118675773581622755375337172702839208422 /usr/lib/python3.11/site-packages/sage/misc/package_dir.py
**********************************************************************
File "/usr/lib/python3.11/site-packages/sage/misc/package_dir.py", line 110, in sage.misc.package_dir.read_distribution
Failed example:
    read_distribution(os.path.join(SAGE_SRC, 'sage', 'graphs', 'graph_decompositions', 'tdlib.pyx'))
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python3.11/site-packages/sage/doctest/forker.py", line 709, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/usr/lib/python3.11/site-packages/sage/doctest/forker.py", line 1144, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.misc.package_dir.read_distribution[2]>", line 1, in <module>
        read_distribution(os.path.join(SAGE_SRC, 'sage', 'graphs', 'graph_decompositions', 'tdlib.pyx'))
      File "/usr/lib/python3.11/site-packages/sage/misc/package_dir.py", line 116, in read_distribution
        with open_source_file(src_file, error_handling='ignore') as fh:
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/python3.11/site-packages/Cython/Utils.py", line 255, in open_source_file
        raise FileNotFoundError(source_filename)
    FileNotFoundError: /usr/lib/python3.11/site-packages/sage/graphs/graph_decompositions/tdlib.pyx
```
prior to this PR, the test were fine because the files concerned would be just shipped as part of sagemath-standard. So far we have dealt with doctests requiring real access to the sources with `#optional - sage_spkg`. Should we do mark those three in the same way? Or is it time we think we think about something more formal when running `sage -t --installed`?

**Pull request merged**